src/threads.rs

```diff
@@ -0,0 +1,303 @@
+//! Implements threads.
+
+use std::cell::RefCell;
+use std::collections::hash_map::Entry;
+
+use log::trace;
+
+use rustc_middle::ty;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_index::vec::{Idx, IndexVec};
+
+use crate::*;
+
+/// A thread identifier.
+#[derive(Clone, Copy, Debug, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct ThreadId(usize);
+
+impl Idx for ThreadId {
+    fn new(idx: usize) -> Self {
+        ThreadId(idx)
+    }
+    fn index(self) -> usize {
+        self.0
+    }
+}
+
+impl From<u64> for ThreadId {
+    fn from(id: u64) -> Self {
+        Self(id as usize)
+    }
+}
+
+/// The state of a thread.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum ThreadState {
+    /// The thread is enabled and can be executed.
+    Enabled,
+    /// The thread tried to join the specified thread and is blocked until that
+    /// thread terminates.
+    Blocked(ThreadId),
+    /// The thread has terminated its execution (we do not delete terminated
+    /// threads.)
+    Terminated,
+}
+
+/// A thread.
+pub struct Thread<'mir, 'tcx> {
+    state: ThreadState,
+    /// The virtual call stack.
+    stack: Vec<Frame<'mir, 'tcx, Tag, FrameData<'tcx>>>,
+    /// Is the thread detached?
+    ///
+    /// A thread is detached if its join handle was destroyed and no other
+    /// thread can join it.
+    detached: bool,
```

Comment:
When a thread is still attached, it will exit when it's parent exits, right? By the way is such a cancel considered UB?

Reply:
> When a thread is still attached, it will exit when it's parent exits, right?

@bjorn3 I was not aware that threads should have such behaviour. Do you have a link to the documentation?

Reply:
@bjorn3 I would be interested in that as well. :)

Reply:
Discussion on Zulip indicates that this is probably not the case -- in fact, the parent relation seems to be entirely irrelevant for thread behavior.